gitpod-io/gitpod
# Migrate account attribution to team attribution

## Description
The logic for migrating user accounts to teams, hidden behind the feature flag`team_only_attribution`.

Checks if users should be migrated based on whether one of the following is true:
 - are not a member of any team
 - has not been migrated (flag in user)

The migration does the following changes:
- create a team and make the user the single member. The name is the user's fullName || name. On conflict the logic adds a number and fails after 10 tries.
- move all projects associated with the user to the new team
- attributes all workspace instances attributed to the user to the new team.
- updates all references of the user's attributionId to the new team (d_b_usage, d_b_cost_center, d_b_stripe_customer).

## Related Issue(s)

See #15665

## How to test

 - login into preview env
 - start a workspace on any repository
 - create a project under the individual account by going to /new
 - create a workspace on that project
 - verify that usage exists under /usage (can take a bit as it is updated by a control loop)
 - (optional) sign-up with stripe

#### trigger migration
- change the user's email to `toa@<somedomain>`
- reload the page and verify that all usage, workspaces and projects have been migrated to a newly created team
- check the usage tab of that team and verify it correctly shows the usage. Check the billing page under settings. It should show the exact same subscription the individual had before.
- check billing in the user settings and verify that you can neither select the individual account nor can you subscribe to anything there. Also no more usage view on the individual account

## Release Notes

```release-note
NONE
```

## Documentation


## Werft options:

      If enabled this will build `install/preview`
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
